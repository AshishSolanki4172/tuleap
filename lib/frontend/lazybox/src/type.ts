/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { HTMLTemplateResult, TemplateResult } from "lit/html.js";
import type { GroupCollection, LazyboxItem } from "./items/GroupCollection";
import type { HTMLTemplateStringProcessor } from "./index";

export interface Lazybox {
    setDropdownContent: (groups: GroupCollection) => void;
    resetSelection: () => void;
    setSelection: (selection: ReadonlyArray<LazyboxItem>) => void;
    destroy: () => void;
}

export type LazyboxSearchFieldCallback = (query: string) => void;

export type LazyboxSelectionCallback = (selected_value: unknown | null) => void;

export type LazyboxTemplatingCallback = (
    html: typeof HTMLTemplateStringProcessor,
    item: LazyboxItem
) => HTMLTemplateResult;

export type LazyboxNewItemCallback = () => void;

export type LazyBoxWithNewItemButton = {
    readonly new_item_button_label: string;
    readonly new_item_callback: LazyboxNewItemCallback;
};

type LazyBoxWithoutNewItemButton = {
    readonly new_item_callback?: undefined;
};

export type LazyboxOptions = (LazyBoxWithoutNewItemButton | LazyBoxWithNewItemButton) & {
    readonly placeholder: string;
    readonly search_input_placeholder: string;
    readonly is_multiple: boolean;
    readonly templating_callback: LazyboxTemplatingCallback;
    readonly selection_callback: LazyboxSelectionCallback;
    readonly search_field_callback: LazyboxSearchFieldCallback;
};

export type RenderedItemMap = Map<string, RenderedItem>;

export interface RenderedItem {
    id: string;
    template: TemplateResult;
    value: unknown;
    is_disabled: boolean;
    is_selected: boolean;
    group_id: string;
    element: Element;
}

export interface LazyboxComponent {
    wrapper_element: HTMLElement;
    lazybox_element: Element;
    dropdown_element: HTMLElement;
    selection_element: HTMLElement;
    placeholder_element: Element;
    dropdown_list_element: HTMLElement;
    search_field_element: HTMLInputElement;
}

export interface LazyboxSelectionStateSingle {
    selected_item: RenderedItem;
    selected_value_element: DocumentFragment;
}

export interface LazyboxSelectionStateMultiple {
    selected_items: Map<string, RenderedItem>;
    selected_values_elements: Map<string, Element>;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}

export interface ManageSelection {
    processSelection(item: Element): void;
    hasSelection(): boolean;
    updateSelectionAfterDropdownContentChange(): void;
    clearSelection(): void;
    setSelection(selection: ReadonlyArray<RenderedItem>): void;
}
