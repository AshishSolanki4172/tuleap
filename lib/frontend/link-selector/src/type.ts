/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { TemplateResult } from "lit/html.js";
import type { LinkSelectorSearchFieldCallback } from "./events/SearchFieldEventCallbackHandler";

export interface LinkSelector {
    destroy: () => void;
}

export interface LinkSelectorOptions {
    readonly placeholder?: string;
    search_field_callback: LinkSelectorSearchFieldCallback;
}

export type RenderedItemMap = Map<string, RenderedItem>;

export interface RenderedItem {
    id: string;
    template: TemplateResult;
    value: string;
    is_disabled: boolean;
    is_selected: boolean;
    group_id: string;
    element: Element;
    target_option: HTMLOptionElement;
}

export interface LinkSelectorComponent {
    wrapper_element: HTMLElement;
    link_selector_element: Element;
    dropdown_element: HTMLElement;
    selection_element: HTMLElement;
    placeholder_element: Element;
    dropdown_list_element: HTMLElement;
    search_field_element: HTMLInputElement;
}

export interface LinkSelectorSelectionStateSingle {
    selected_item: RenderedItem;
    selected_value_element: DocumentFragment;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}
