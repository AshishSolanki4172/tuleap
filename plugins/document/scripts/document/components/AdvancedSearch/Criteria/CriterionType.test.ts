/**
 * Copyright (c) Enalean, 2022 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { shallowMount } from "@vue/test-utils";
import CriterionType from "./CriterionType.vue";
import { createStoreMock } from "@tuleap/core/scripts/vue-components/store-wrapper-jest";
import localVue from "../../../helpers/local-vue";
import type { ConfigurationState } from "../../../store/configuration";

describe("CriterionType", () => {
    it("should render the component", () => {
        const wrapper = shallowMount(CriterionType, {
            localVue,
            propsData: {
                value: "folder",
            },
            mocks: {
                $store: createStoreMock({
                    state: {
                        configuration: {
                            user_can_create_wiki: true,
                        } as ConfigurationState,
                    },
                }),
            },
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it("should omit wiki option if there is no wiki activated in the project", () => {
        const wrapper = shallowMount(CriterionType, {
            localVue,
            propsData: {
                value: "folder",
            },
            mocks: {
                $store: createStoreMock({
                    state: {
                        configuration: {
                            user_can_create_wiki: false,
                        } as ConfigurationState,
                    },
                }),
            },
        });

        expect(wrapper.find("[data-test=type-wiki]").exists()).toBe(false);
    });

    it("should warn parent component when user is changing selection", () => {
        const wrapper = shallowMount(CriterionType, {
            localVue,
            propsData: {
                value: "wiki",
            },
            mocks: {
                $store: createStoreMock({
                    state: {
                        configuration: {
                            user_can_create_wiki: true,
                        } as ConfigurationState,
                    },
                }),
            },
        });

        wrapper.find("[data-test=type-folder]").setSelected();
        expect(wrapper.emitted().input).toStrictEqual([["folder"]]);
    });
});
