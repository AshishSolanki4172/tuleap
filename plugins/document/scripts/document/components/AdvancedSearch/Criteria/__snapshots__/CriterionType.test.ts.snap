// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CriterionType should render the component 1`] = `
<div
  class="tlp-form-element document-search-criterion"
>
  <label
    class="tlp-label"
    data-current-language="en_US"
    data-msgid="Type"
    for="document-type-search"
  >
    Type
  </label>
   
  <select
    class="tlp-select"
    data-test="type"
    id="document-type-search"
  >
    <option
      data-current-language="en_US"
      data-msgid="Any"
      value=""
    >
      Any
    </option>
     
    <option
      data-current-language="en_US"
      data-msgid="
            Folder
        "
      data-test="type-folder"
      value="folder"
    >
      
            Folder
        
    </option>
     
    <option
      data-current-language="en_US"
      data-msgid="File"
      value="file"
    >
      File
    </option>
     
    <option
      data-current-language="en_US"
      data-msgid="Link"
      value="link"
    >
      Link
    </option>
     
    <option
      data-current-language="en_US"
      data-msgid="
            Embedded file
        "
      value="embedded"
    >
      
            Embedded file
        
    </option>
     
    <option
      data-test="type-wiki"
      value="wiki"
    >
      <translate-stub
        tag="span"
      >
        Wiki page
      </translate-stub>
    </option>
     
    <option
      data-current-language="en_US"
      data-msgid="
            Empty document
        "
      value="empty"
    >
      
            Empty document
        
    </option>
  </select>
</div>
`;
